modules/apps/static/portal-lpkg-deployer/portal-lpkg-deployer-impl/src/test/java/com/liferay/portal/lpkg/deployer/internal/ContainerLPKGUtilTest.java

```diff
@@ -0,0 +1,158 @@
+import com.liferay.portal.kernel.util.FileUtil;
+import com.liferay.portal.kernel.util.StreamUtil;
+import com.liferay.portal.lpkg.deployer.test.util.LPKGTestUtil;
+import com.liferay.portal.test.log.LogCapture;
+import com.liferay.portal.test.log.LogEntry;
+import com.liferay.portal.test.log.LoggerTestUtil;
+import com.liferay.portal.test.rule.LiferayUnitTestRule;
+import com.liferay.portal.util.FileImpl;
+import com.liferay.portal.util.PropsValues;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * @author Jiefeng Wu
+ */
+public class ContainerLPKGUtilTest {
+
+	@ClassRule
+	@Rule
+	public static final LiferayUnitTestRule liferayUnitTestRule =
+		LiferayUnitTestRule.INSTANCE;
+
+	@Before
+	public void setUp() throws IOException {
+		Path tempPath = Files.createTempDirectory("testDeploy");
+
+		ReflectionTestUtil.setFieldValue(
+			PropsValues.class, "MODULE_FRAMEWORK_MARKETPLACE_DIR",
+			tempPath.toString());
+
+		ReflectionTestUtil.setFieldValue(
+			FileUtil.class, "_file", FileImpl.getInstance());
+
+		_tempFolder = tempPath.toFile();
+	}
+
+	@After
+	public void tearDown() {
+		FileUtil.deltree(_tempFolder);
+	}
+
+	@NewEnv(type = NewEnv.Type.CLASSLOADER)
+	@Test
+	public void testDeploy() throws Exception {
+		Assert.assertEquals(
+			_createExpectedOutputFileList("inner.lpkg"),
+			ContainerLPKGUtil.deploy(
+				_createLPKGContainerFile("inner.lpkg"), null));
+
+		Assert.assertTrue(FileUtil.exists(new File(_tempFolder, "inner.lpkg")));
+	}
+
+	@NewEnv(type = NewEnv.Type.CLASSLOADER)
+	@Test
+	public void testDeployWithZipSlip() throws Exception {
+		try (LogCapture logCapture = LoggerTestUtil.configureJDKLogger(
+				ContainerLPKGUtil.class.getName(), Level.WARNING)) {
+
+			Assert.assertEquals(
+				_createExpectedOutputFileList("good.lpkg"),
+				ContainerLPKGUtil.deploy(
+					_createLPKGContainerFile("good.lpkg", "../bad.lpkg"),
+					null));
+
+			Assert.assertTrue(
+				FileUtil.exists(new File(_tempFolder, "good.lpkg")));
+
+			Assert.assertFalse(
+				FileUtil.exists(new File(_tempFolder.getParent(), "bad.lpkg")));
+
+			List<LogEntry> logEntries = logCapture.getLogEntries();
+
+			Assert.assertEquals(logEntries.toString(), 1, logEntries.size());
+
+			LogEntry logEntry = logEntries.get(0);
+
+			Assert.assertEquals(
+				"Invalid LPKG File name: ../bad.lpkg", logEntry.getMessage());
+		}
+	}
+
+	private List<File> _createExpectedOutputFileList(String fileName) {
```

Comment:
@jeffwu0724 Please inline this method. It is only one line, do not need to be a method.